### PR TITLE
Revs now have a delay before starting

### PR DESCRIPTION
**Pull request by coolmankid12345**
## About the PR

Revs now start somewhere between 2 to 5 minutes after the shift starts. (Don't fill your bags)
## Why / Balance

Slightly so it's more in line with the other antags and mostly so it gives time for late joiners to still have a chance to get head rev. Also so you don't spawn in and instantly get converted by the guy next to you.
## Technical details

ActiveTick now checks when the game should start and if Revs were already chosen or not.
## Media

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

### CODE REVIEW

**Comment by JoeHammad1844** [comment]
there could definitely be some conflict with bag fills, what happens if you dont have a bag or its full?

**Comment by chromiumboy** [comment]
> there could definitely be some conflict with bag fills, what happens if you dont have a bag or its full?

This. Having a flash spawn in you hands, or next to you, could instantly out you depending on the circumstances

**Comment by coolmankid12345** [comment]
> there could definitely be some conflict with bag fills, what happens if you dont have a bag or its full?

> This. Having a flash spawn in you hands, or next to you, could instantly out you depending on the circumstances

Currently, it tries bag and then pockets, if neither of those work it gets stuck to your player and you take the L. I could add hands too. Hell I could just force it into the bag regardless of space but I doubt that's the best course of action cause then you end up with a fucked up bag.

**Comment by JoeHammad1844** [comment]
I don't think its worth it for the.. immersion? its just gonna be a burden

**Comment by Nairodian** [comment]
Make it a button in the PDA that dispenses the flash into your hand and then the button disappears. Basically how syndicates work-ish, kind of, a bit. This seems like the most sane way though, but could be hard, I don't know, I don't code much.

**Comment by Tunguso4ka** [comment]
> Make it a button in the PDA

uplink for the revs, with 3tc in. 1 tc for sunglasses, 2 for flash.

**Comment by JoeHammad1844** [comment]
goonstation did that, heres what they can buy 
image

**Comment by Endecc** [comment]
Could skip over the uplink entirely, maybe. Give headrevs an uplink code when it's time to roll, and slot them their items when they input it to their PDA.

Combining having a discreet way to get your equipment, without strangely reusing Traitor uplinks for only two items.

**Comment by JoeHammad1844** [comment]
how about they get a headrev uplink implant and they can get their initial sunglasses and flash for free and the rest is paid? also rev flashbangs would be sick

**Comment by JoeHammad1844** [comment]
you see all this crazy stuff we're thinking of? this is why headrevs need to be active roundstart, its just gonna be too complicated to have them be chosen later and recieve gear

**Comment by chromiumboy** [comment]
> I disagree and feel like this stance makes absolutely no sense. It is insanely and **unfairly** easy to insta-rev everyone in your department on round start because you all spawn next to each other and no one is expecting to be flashed. There is no counter play. This is something I've seen several times over while admining. Syndicates already have a delay and have WAY more gear than headrevs, yet do just fine with the delay.

Head revs can do this at anytime though, it's not limited to round start. It's a very risky move to make too. All you need is one person in the group, or in an adjacent room, to type out 'REVS' on the radio and you are screwed

There isn't really much counterplay to revs in general, as flash protection is largely limited to sec, and they can't be converted. Personally I think we should move away from using flashes to convert people

**Comment by JoeHammad1844** [comment]
> I disagree and feel like this stance makes absolutely no sense. It is insanely and **unfairly** easy to insta-rev everyone in your department on round start because you all spawn next to each other and no one is expecting to be flashed. There is no counter play. This is something I've seen several times over while admining. Syndicates already have a delay and have WAY more gear than headrevs, yet do just fine with the delay.

traitors dont get roundstart gear though, they buy it, my issue is that the rev gear could just spawn on the floor and give them away

and converting people round start is very high risk high reward, and they should be rewarded for pulling it off before they can scream on radio. its also hard to do, especially in places like aspid salvage

**Comment by JoeHammad1844** [comment]
>  I think we should move away from using flashes to convert people

well its worked for literally every single ss13 server that has revs, and how else could they be converted? some kindof injection is too long and would require stunning which is basically unobtanium unless you're sec

**Comment by chromiumboy** [comment]
> > I think we should move away from using flashes to convert people
> 
> well its worked for literally every single ss13 server that has revs, and how else could they be converted? some kindof injection is too long and would require stunning which is basically unobtanium unless you're sec

One idea I had is that head revs have a hypnosis ability. When used on an unprotected target, they are converted to the cause. This could be instant, like with a flash, but I think having a short duration would be better - the target would be paralyzed, unable to move or speak while the conversion is in progress

There's a couple reasons I'd like this over using flashes. You can balance rev conversions independently of how flashes function, as well as how hard flashes are to find / manufacture. Revs wouldn't be instantly outted by finding flashes and flash protections on someone. Also, by having a duration on conversion, revs can't convert multiple people in quick succession, and converting someone with someone else present would be a big risk (as damage would interupt the conversion). This would encourage head revs to be sneaky and pick people off one by one. You can also balance the total number of revs that a single head can convert in a round by having a limited number of charges on their conversion ability (say between 5 and 10), and even potentially add interesting ways of gaining new charges for their ability (such as completing syndicate like objectives) rather than just printing / stealing more flashes

**Comment by JoeHammad1844** [comment]
> There isn't really much counterplay to revs in general, as flash protection is largely limited to sec

security can print sunglasses at their techfab, they have two in the sectec and a pair in each secoff locker, this is a non issue if sec knows what they are doing

File: Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs [hunk 1]
```diff
@@ -137,27 +161,15 @@ private void OnStartAttempt(RoundStartAttemptEvent ev)
...
-
     private void GiveHeadRev(List<EntityUid> chosen, string antagProto, RevolutionaryRuleComponent comp)
     {
+        comp.HeadRevsChosen = true;
         foreach (var headRev in chosen)
         {
             RemComp<CommandStaffComponent>(headRev);
+            //In case some funny guy mindshields themselves before being selected
+            if (TryComp<MindShieldComponent>(headRev, out var mindshield))
+                RemComp<MindShieldComponent>(headRev);
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs:172]: RemComp doesn't throw if it doesn't exist.

File: Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs [hunk 2]
```diff
@@ -74,14 +76,36 @@ protected override void Started(EntityUid uid, RevolutionaryRuleComponent compon
...
-                GameTicker.EndGameRule(uid, gameRule);
+                _antagSelection.EligiblePlayers(comp.HeadRevPrototypeId, comp.MaxHeadRevs, comp.PlayersPerHeadRev, comp.HeadRevStartSound,
+                    "head-rev-role-greeting", "#5e9cff", out var chosen);
+                if (chosen.Any())
+                    GiveHeadRev(chosen, comp.HeadRevPrototypeId, comp);
+                else
+                {
+                    _chatManager.SendAdminAnnouncement(Loc.GetString("rev-no-heads"));
+                    comp.HeadRevsChosen = true;
+                }
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs:108]: Extremely weird code flow where you set headrevs outside the method but also inside the method.

Also you can just remove headrevschosen and just use the start time anyway as they are both functioning similarly.

File: Content.Server/GameTicking/Rules/Components/RevolutionaryRuleComponent.cs [hunk 3]
```diff
@@ -59,6 +59,30 @@ public sealed partial class RevolutionaryRuleComponent : Component
...
+    /// Whether or not Head Revs have been selected yet or not.
+    /// </summary>
+    [DataField]
+    public bool HeadRevsChosen;
+
+    /// <summary>
+    /// When Revs will start during a shift.
+    /// </summary>
+    [DataField(customTypeSerializer: typeof(TimeOffsetSerializer)), ViewVariables(VVAccess.ReadWrite)]
+    public TimeSpan? StartTime;
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/Components/RevolutionaryRuleComponent.cs:72]: Needs offsetting on unpause.

File: Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs (hunk 2 above)
Comment (coolmankid12345) [reply Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs]: I have headrevschosen so this antag selection code doesn't trigger repeatedly as it's used above at the command check with a return. If I had both on start time I think only the command check would trigger as it wouldn't reach the antag selection.

File: Content.Server/GameTicking/Rules/Components/RevolutionaryRuleComponent.cs (hunk 3 above)
Comment (coolmankid12345) [reply Content.Server/GameTicking/Rules/Components/RevolutionaryRuleComponent.cs:72]: I'm confused, isn't that what timeoffsetserializer is for or am I mistaken?

Comment (DrSmugleaf) [reply Content.Server/GameTicking/Rules/Components/RevolutionaryRuleComponent.cs:72]: The serializer is for applying it when the component is read or copied, but you also need to add a subscription for EntityUnpausedEvent that increments the start time by the time that the rule entity was paused for.
Find-in-files `EntityUnpausedEvent>` should give you some examples to look at, such as the one in `AnomalySystem.Generator.cs`

**Comment by Emisse** [comment]
lg2m

**Review by Emisse — APPROVED** [review]

File: Content.Server/GameTicking/Rules/Components/RevolutionaryRuleComponent.cs [hunk 4]
```diff
@@ -59,6 +59,30 @@ public sealed partial class RevolutionaryRuleComponent : Component
...
+    /// </summary>
+    [DataField]
+    public TimeSpan MinStartDelay = TimeSpan.FromMinutes(2);
+
+    /// <summary>
+    /// The maximum amount of time after the round starts that the Head Revs will be chosen.
+    /// </summary>
+    [DataField]
+    public TimeSpan MaxStartDelay = TimeSpan.FromMinutes(5);
+
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/Components/RevolutionaryRuleComponent.cs:85]: Station rules already have fields for these.

File: Content.Server/Antag/AntagSelectionSystem.cs [hunk 5]
```diff
@@ -275,27 +276,41 @@ public void GiveAntagBagGear(EntityUid antag, string item)
...
+            else if (_inventory.TryGetSlotContainer(antag, "pocket2", out var pocket2Slot, out _) && pocket2Slot.ContainedEntity == null && _containerSystem.CanInsert(itemToSpawn, pocket2Slot))
+            {
+                pocket2Slot.Insert(itemToSpawn);
+                return;
+            }
+            TryHands(antag, itemToSpawn);
+        }
+        else
+            TryHands(antag, itemToSpawn);
+    }
```
Comment (metalgearsloth) [inline Content.Server/Antag/AntagSelectionSystem.cs:297]: TryInsert

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
This looks like it's just copying zombie code to me.

Realistically since revs are disabled I'm not going to waste your time by requesting changes.

**Pull request closed without merging**